### PR TITLE
ARROW-7420: [C++] Migrate tensor related APIs to Result-returning version

### PR DESCRIPTION
**Pull request by mrkn**
I'd like to modify tensor-related APIs, both internal and public, to use `Result<T>` for the return values.

Moreover, I've fixed a bug of `SparseCSXIndex::ValidateShape` for CSC format.

### CODE REVIEW

File: cpp/src/arrow/sparse_tensor.cc [hunk 1]
```diff
@@ -33,7 +33,7 @@ namespace arrow {
 // SparseIndex
 
 Status SparseIndex::ValidateShape(const std::vector<int64_t>& shape) const {
-  if (!std::all_of(shape.begin(), shape.end(), [](int64_t x) { return x > 0; })) {
+  if (!std::all_of(shape.begin(), shape.end(), [](int64_t x) { return x >= 0; })) {
```
Comment (mrkn) [inline cpp/src/arrow/sparse_tensor.cc]: This and the same change in tensor.c below is necessary to accept making tensors with 0-length dimensions at https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/read_write_test.cc#L1112.

File: cpp/src/arrow/ipc/reader.cc [hunk 2]
```diff
@@ -1038,16 +1028,15 @@ Status CheckSparseTensorBodyBufferCount(
 
 }  // namespace
 
-Status ReadSparseTensorBodyBufferCount(const Buffer& metadata, size_t* buffer_count) {
+Result<size_t> ReadSparseTensorBodyBufferCount(const Buffer& metadata) {
   SparseTensorFormat::type format_id;
 
   RETURN_NOT_OK(internal::GetSparseTensorMetadata(metadata, nullptr, nullptr, nullptr,
```
Comment (fsaintjacques) [inline cpp/src/arrow/ipc/reader.cc]: Seems like you can also refactor `GetSparceTensorMetadata` since it's internal and Tensor?

Comment (mrkn) [reply cpp/src/arrow/ipc/reader.cc]: `GetSparseTensorMetadata` has 5 output variables, so I think it is good to remain that this function returns a `Status`.

Or, should we use tuple for that case?

**Review by wesm — APPROVED** [review]
+1. I'm a bit about ambivalent about refactoring internal APIs (though it's good for cleanliness long term) since the public APIs are what matter from a longer-term / stability standpoint

**Pull request closed without merging**